ranger/config/commands.py

```diff
@@ -1212,6 +1212,100 @@ def execute(self):
             fm.notify("files have not been retagged")
 
 
+class bulk(Command):
+    """:bulk <attribute>
+
+    This command opens a list of with the attribute <attribute> for each of the
+    selected files in an external editor.
+    After you edit and save the file, it will generate a shell script
+    which changes the attributes in bulk according to the changes you did in the
+    file.
+
+    This shell script is opened in an editor for you to review.
+    After you close it, it will be executed.
+    """
+
+
+    class id3art(object):
+
+        def get_attribute(self, file):
+            import eyed3
+            return str(eyed3.load(file.relative_path).tag.artist)
+
+
+        def get_change_attribute_cmd(self, file, old, new):
+            from ranger.ext.shell_escape import shell_escape as esc
+            return "eyeD3 -a %s %s" % (esc(new), esc(file))
+
+    class id3tit(object):
+
+        def get_attribute(self, file):
+            import eyed3
+            return str(eyed3.load(file.relative_path).tag.title)
+
+
+        def get_change_attribute_cmd(self, file, old, new):
+            from ranger.ext.shell_escape import shell_escape as esc
+            return "eyeD3 -t %s %s" % (esc(new), esc(file))
+
+    bulk = {'id3art': id3art(),
```

Comment:
Hmm, would be nice if this worked more like image preview methods, where you can register a method. That way this code wouldn't have to be copied and edited but people could you define a class and register.

Reply:
Indeed it is better. I too found using this dictionary cumbersome, but didn't know which better interface to use.

I now have added a new ext/bulk.py file containing the same interface used in the image preview methods but for these bulk commands. I was going to add it all inside the bulk(Command) class but decorator definition doesn't seem to work inside classes, or I'm doing something wrong. But it looked a little too chaotic anyway, so I moved everything to the new file (except for the commands).

I also made the `id3_artist` and `id3_title` bulk commands extend a base `BulkCommand` class. That way users can use the base class as reference for which functions need to be implemented and which are the parameters.